presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java

```diff
@@ -274,10 +270,57 @@ public QueryResultRows build()
         {
             return new QueryResultRows(
                     session,
-                    types,
                     columns,
                     pages.build(),
                     exceptionConsumer);
         }
+
+        private static Optional<List<ColumnAndType>> combine(List<Column> columns, List<Type> types)
+        {
+            if (columns == null || types == null) {
+                return Optional.empty();
+            }
+
+            verify(columns.size() == types.size(), "columns and types size mismatch");
+
+            ImmutableList.Builder<ColumnAndType> builder = ImmutableList.builder();
+
+            for (int i = 0; i < columns.size(); i++) {
+                builder.add(new ColumnAndType(columns.get(i), types.get(i)));
+            }
+
+            return Optional.of(builder.build());
+        }
+    }
+
+    private static class ColumnAndType
+    {
+        private final Column column;
+        private final Type type;
```

Comment:
That's a bit unintuitive because `Column` also has type information (`String type` and `ClientTypeSignature typeSignature`)

cc @electrum ?

Reply:
I followed logic of previous implementation. Columns are types are collected in `setQueryOutputInfo` method:

```
private synchronized void setQueryOutputInfo(QueryExecution.QueryOutputInfo outputInfo)
    {
        // if first callback, set column names
        if (columns == null) {
            List<String> columnNames = outputInfo.getColumnNames();
            List<Type> columnTypes = outputInfo.getColumnTypes();
            checkArgument(columnNames.size() == columnTypes.size(), "Column names and types size mismatch");

            ImmutableList.Builder<Column> list = ImmutableList.builder();
            for (int i = 0; i < columnNames.size(); i++) {
                list.add(createColumn(columnNames.get(i), columnTypes.get(i)));
            }
            columns = list.build();
            types = outputInfo.getColumnTypes();
        }

        for (URI outputLocation : outputInfo.getBufferLocations()) {
            exchangeClient.addLocation(outputLocation);
        }
        if (outputInfo.isNoMoreBufferLocations()) {
            exchangeClient.noMoreLocations();
        }
    }
```